backend/src/od.rs

```diff
@@ -0,0 +1,148 @@
+                        .map(|obj| obj.data),
+                ) {
+                    if i1 != i2 {
+                        requests.push((i1, i2, trip_count));
+                    }
+                }
+            }
+        }
+        requests
+    }
+
+    pub fn to_gj(&self, map: &MapModel) -> GeoJson {
+        // Per (from, to) pair, how many trips?
+        let mut from: Vec<Vec<usize>> =
+            std::iter::repeat_with(|| std::iter::repeat(0).take(self.zones.len()).collect())
+                .take(self.zones.len())
+                .collect();
+        // Per (to, from) pair, how many trips?
+        let mut to: Vec<Vec<usize>> =
+            std::iter::repeat_with(|| std::iter::repeat(0).take(self.zones.len()).collect())
+                .take(self.zones.len())
+                .collect();
+
+        for (zone1, zone2, count) in &self.desire_lines {
+            from[zone1.0][zone2.0] += count;
+            to[zone2.0][zone1.0] += count;
+        }
+
+        let mut features = Vec::new();
+        for (idx, zone) in self.zones.iter().enumerate() {
+            let mut f = map.mercator.to_wgs84_gj(&zone.geometry);
+            f.set_property("name", zone.name.clone());
+            f.set_property("counts_from", std::mem::take(&mut from[idx]));
+            f.set_property("counts_to", std::mem::take(&mut to[idx]));
+            features.push(f);
+        }
+        GeoJson::from(features)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Zone {
+    /// An original opaque string ID, from different data sources.
+    pub name: String,
+    /// WGS84 when built originally and serialized, then Mercator right before being used
+    pub geometry: MultiPolygon,
+    /// The bbox, rounded to centimeters, for generate_range to work. Only calculated right before
+    /// use.
+    #[serde(skip_deserializing, skip_serializing)]
+    pub x1: i64,
+    #[serde(skip_deserializing, skip_serializing)]
+    pub y1: i64,
+    #[serde(skip_deserializing, skip_serializing)]
+    pub x2: i64,
+    #[serde(skip_deserializing, skip_serializing)]
+    pub y2: i64,
+}
+
+impl Zone {
+    fn random_point(&self, rng: &mut WyRand) -> Point {
+        loop {
+            let x = (rng.generate_range(self.x1..=self.x2) as f64) / 100.0;
+            let y = (rng.generate_range(self.y1..=self.y2) as f64) / 100.0;
+            let pt = Point::new(x, y);
+            if self.geometry.contains(&pt) {
+                return pt;
+            }
+        }
+    }
+}
+
+/// Deterministically produce a bunch of OD pairs, just as a fallback when there's no real data
+pub fn synthetic_od_requests(map: &MapModel) -> Vec<(IntersectionID, IntersectionID, usize)> {
+    let num_requests = 1_000;
+
+    let mut rng = WyRand::new_seed(42);
+    let mut requests = Vec::new();
+    for _ in 0..num_requests {
+        let i1 = IntersectionID(rng.generate_range(0..map.intersections.len()));
+        let i2 = IntersectionID(rng.generate_range(0..map.intersections.len()));
```

Comment:
Should we check that i1 != i2?

Reply:
Done in c6b5c67ad47b905ae3697c538baf0861f2e0c273